File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string, int> player_indices;
    for (int i = 0; i < players.size(); ++i) {
        player_indices[players[i]] = i;
    }

    for (auto& calling : callings) {
        int idx = player_indices[calling];
        if (idx > 0) {
            swap(players[idx - 1], players[idx]);
            player_indices[players[idx]] = idx;
            player_indices[players[idx - 1]] = idx - 1;
        }
    }
    
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.*;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        
        Map<String, Integer> playerRank = new HashMap<>();
        Map<Integer, String> rankPlayer = new HashMap<>();
        
        for(int i = 0 ; i < players.length ; i++){
            playerRank.put(players[i], i);
            rankPlayer.put(i, players[i]);
        }
        
        for(String calling : callings) {
            
            int curRank = playerRank.get(calling);
            int newRank = curRank - 1;
            String prevPlayer = rankPlayer.get(newRank);
            
            rankPlayer.put(newRank, calling);
            rankPlayer.put(curRank, prevPlayer);
            
            playerRank.put(calling, newRank);
            playerRank.put(prevPlayer, curRank);      
            
        }
        
        List<String> result = new ArrayList<>();
        for(int i = 0 ; i < players.length ; i++)
            result.add(rankPlayer.get(i));
        
        return result.stream().toArray(String[]::new);
        
    }
}